interp/memory.go

```diff
@@ -0,0 +1,1392 @@
+	if mv.objects == nil {
+		mv.objects = make(map[uint32]object)
+	}
+	if checks && mv.get(index).buffer == nil {
+		panic("writing to external object")
+	}
+	if checks && mv.get(index).buffer.len(mv.r) != obj.buffer.len(mv.r) {
+		panic("put() with a differently-sized object")
+	}
+	mv.objects[index] = obj
+}
+
+// Load the value behind the given pointer.
+func (mv *memoryView) load(p pointerValue, size uint32) value {
+	if checks && mv.hasExternalStore(p) {
+		panic("interp: load from object with external store")
+	}
+	obj := mv.get(p.index())
+	if p.offset() == 0 && size == obj.size {
+		return obj.buffer.clone()
+	}
+	if checks && p.offset()+size > obj.size {
+		panic("interp: load out of bounds")
+	}
+	v := obj.buffer.asRawValue(mv.r)
+	loadedValue := rawValue{
+		buf: v.buf[p.offset() : p.offset()+size],
+	}
+	return loadedValue
+}
+
+// Store to the value behind the given pointer. This overwrites the value in the
+// memory view, so that the changed value is discarded when the memory view is
+// reverted.
+func (mv *memoryView) store(v value, p pointerValue) {
+	if checks && mv.hasExternalLoadOrStore(p) {
+		panic("interp: store to object with external load/store")
+	}
+	obj := mv.get(p.index())
+	if checks && p.offset()+v.len(mv.r) > obj.size {
+		panic("interp: store out of bounds")
+	}
+	if p.offset() == 0 && v.len(mv.r) == obj.buffer.len(mv.r) {
+		obj.buffer = v
+	} else {
+		obj = obj.clone()
+		buffer := obj.buffer.asRawValue(mv.r)
+		obj.buffer = buffer
+		v := v.asRawValue(mv.r)
+		for i := uint32(0); i < v.len(mv.r); i++ {
+			buffer.buf[p.offset()+i] = v.buf[i]
+		}
+	}
+	mv.put(p.index(), obj)
+}
+
+// value is some sort of value, comparable to a LLVM constant. It can be
+// implemented in various ways for efficiency, but the fallback value (that all
+// implementations can be converted to except for localValue) is rawValue.
+type value interface {
+	// len returns the length in bytes.
+	len(r *runner) uint32
+	clone() value
+	asPointer(*runner) (pointerValue, error)
+	asRawValue(*runner) rawValue
+	Uint() uint64
+	Int() int64
+	toLLVMValue(llvm.Type, *memoryView) llvm.Value
+	String() string
+}
+
+// literalValue contains simple integer values that don't need to be stored in a
+// buffer.
+type literalValue struct {
+	value interface{}
+}
+
+func (v literalValue) len(r *runner) uint32 {
+	switch v.value.(type) {
+	case uint64:
+		return 8
+	case uint32:
+		return 4
+	case uint16:
+		return 2
+	case uint8:
+		return 1
+	default:
+		panic("unknown value type")
+	}
+}
+
+func (v literalValue) String() string {
+	return strconv.FormatInt(v.Int(), 10)
+}
+
+func (v literalValue) clone() value {
+	return v
+}
+
+func (v literalValue) asPointer(r *runner) (pointerValue, error) {
+	return pointerValue{}, errLiteralToPointer
+}
+
+func (v literalValue) asRawValue(r *runner) rawValue {
+	var buf []byte
+	switch value := v.value.(type) {
+	case uint64:
+		buf = make([]byte, 8)
+		binary.LittleEndian.PutUint64(buf, value)
```

Comment:
I don't think we have to deal with any big-endian machines right now, but we should probably document this limitation somewhere.

Reply:
Not sure whether this is necessary. If we ever add support for a big endian system (if they are still around), grepping for `LittleEndian` seems pretty obvious to me. Right now there is at least one other place that assumes little endian systems: builder/build.go (although it's perhaps less important as each architecture needs special code to be supported there).

Reply:
Makes sense to me!

Reply:
I will document this at the top of the file just to be sure.